app/services/hashed_password.rb

```diff
@@ -0,0 +1,75 @@
+# HashedPassword is returned from PasswordHasher.hash.
+# HashedPassword outputs various key hashing related to NIST encryption.
+#
+# Generate and store a 128-bit salt S.
+# Z1, Z2 = scrypt(S, password)   # split 256-bit output into two halves
+# Generate random R.
+# D = KMS_GCM_Encrypt(key=server_secret, plaintext=R) xor Z1
+# E = hash( Z2 + R )
+# F = hash(E)
+# Store F (User.encrypted_password) and D (User.encryption_key) in db
+
+class HashedPassword
```

Comment:
My problem is that I think of password as a simple string, and the same with any sort of hash. The object we're calling `HashedPassword` is not a password, it's a class that does things to a password string to me. So `NistPassword` does not help IMO.

And on top of that, `PasswordHasher.hash` ... I'd expect that to return the aforementioned simple string.

What about

```ruby
partial_password_hash = PasswordHasher.partial_hash(password)

partial_password_hash.hash_f
```